tests/UnifiedSpecTests/Constraint/IsBsonType.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+namespace MongoDB\Tests\UnifiedSpecTests\Constraint;
+
+use LogicException;
+use MongoDB\BSON\BinaryInterface;
+use MongoDB\BSON\DBPointer;
+use MongoDB\BSON\Decimal128Interface;
+use MongoDB\BSON\Int64;
+use MongoDB\BSON\JavascriptInterface;
+use MongoDB\BSON\MaxKeyInterface;
+use MongoDB\BSON\MinKeyInterface;
+use MongoDB\BSON\ObjectIdInterface;
+use MongoDB\BSON\RegexInterface;
+use MongoDB\BSON\Symbol;
+use MongoDB\BSON\TimestampInterface;
+use MongoDB\BSON\Undefined;
+use MongoDB\BSON\UTCDateTimeInterface;
+use MongoDB\Model\BSONArray;
+use PHPUnit\Framework\Constraint\Constraint;
+use RuntimeException;
+use Symfony\Bridge\PhpUnit\ConstraintTrait;
+use function array_keys;
+use function count;
+use function is_array;
+use function is_bool;
+use function is_float;
+use function is_int;
+use function is_object;
+use function is_string;
+use function range;
+use function sprintf;
+use const PHP_INT_SIZE;
+
+final class IsBsonType extends Constraint
+{
+    use ConstraintTrait;
+
+    private static $knownTypes = [
+        'double' => 1,
+        'string' => 1,
+        'object' => 1,
+        'array' => 1,
+        'binData' => 1,
+        'undefined' => 1,
+        'objectId' => 1,
+        'bool' => 1,
+        'date' => 1,
+        'null' => 1,
+        'regex' => 1,
+        'dbPointer' => 1,
+        'javascript' => 1,
+        'symbol' => 1,
+        'javascriptWithScope' => 1,
+        'int' => 1,
+        'timestamp' => 1,
+        'long' => 1,
+        'decimal' => 1,
+        'minKey' => 1,
+        'maxKey' => 1,
+    ];
+
+    private $type;
+
+    public function __construct(string $type)
+    {
+        if (! isset(self::$knownTypes[$type])) {
+            throw new RuntimeException(sprintf('Type specified for %s <%s> is not a valid type', self::class, $type));
+        }
+
+        $this->type = $type;
+    }
+
+    private function doMatches($other) : bool
+    {
+        switch ($this->type) {
+            case 'double':
+                return is_float($other);
+            case 'string':
+                return is_string($other);
+            case 'object':
+                return is_object($other) && (! $other instanceof BSONArray);
```

Comment:
Any BSON type that is represented by an object, including a 64-bit integer on a 32-bit platform will return true for this check. Is that intended? I'm wondering if this check should not only exclude `BSONArray`, but also exclude any `MongoDB\BSON\Type` instance to avoid false positives?

Edit: noticed that I built the original code in `DocumentMatchConstraint`, so I'll happily take the blame for missing any type overlaps while writing this.

Reply:
Ah, good catch. I did indeed copy this from DocumentMatchConstraint. It was just the array detection that I rewrote.

I'll break this up into a separate `isObject()` method to cover the cases.

Reply:
This should now be fixed. If my new approach looks good, free to backport to v1.6. IIRC, that's when DocumentMatchConstraint got `$$type` support.